[CLD-21]: feat(operations): introduce idempotency behaviour

graham-chainlink opened the pull request:
When a sequence or an operation is executed, it nows checks for previous execution in the reporter and if a successful execution was found in the past, it will skip the current execution and return the previous successful result.

Execution will not be skipped if the previous execution was a failure.

An operation or sequence uniqueness is determine by its definition (id + version +description) and its input

The current execution report will not include any reports for operations/sequences that were skipped.

JIRA: https://smartcontract-it.atlassian.net/browse/CLD-21

graham-chainlink committed:
feat(operations): introduce idempotency behaviour

When a sequence or an operation is executed, it nows checks for previous execution in the reporter and if a successful execution was found in the past, it will skip the current execution and return the previous successful result.

Execution will not be skipped if the previous execution was a failure.

An operation or sequence uniqueness is determine by its definition (id + version +description) and its input

The current execution report will not include any reports for operations/sequences that were skipped.

JIRA: https://smartcontract-it.atlassian.net/browse/CLD-21

giogam commented on deployment/operations/execute.go:
```diff
@@ -110,25 +149,59 @@ func ExecuteSequence[IN, OUT, DEP any](
+
+	for _, report := range prevReports {
+		// Check if operation/sequence was run previously and return the report if successful
+		reportHash, err := constructUniqueHashFrom(report.Def, report.Input)
```
Is there a way to cache the report hash alongside the report once it’s computed, so we can avoid repeatedly calling constructUniqueHashFrom for the same Def and Input pair?

graham-chainlink replied:
yeah valid point, i could use a faster non-cryptographic hash function but it wont have clash resistance.
So the choice is either slower, but with clash resistance or faster but without clash resistance, just trade off. My gut feeling now is that the number of reports will be quite low as my plan is to separate reports per migration run, so i dont think performance will be an issue here.

and if it is, then we can swap the hashing algo here in the future.

graham-chainlink replied:
Good idea, added in the 2nd commit.

giogam commented on deployment/operations/execute.go:
```diff
@@ -110,25 +149,59 @@ func ExecuteSequence[IN, OUT, DEP any](
+		return "", err
+	}
+
+	hash := sha256.Sum256(append(defBytes, inputBytes...))
```
Mostly a note on potential optimization: if `sha256` turns out to be too computationally expensive, it might be worth exploring faster, perhaps "less secure", hash functions. Even non-cryptographic ones, since strong cryptographic guarantees aren’t strictly required in this context.

graham-chainlink replied:
yeah valid point, i could use a faster non-cryptographic hash function but it wont have clash resistance.
So the choice is either slower, but with clash resistance or faster but without clash resistance, just trade off. My gut feeling now is that the number of reports will be quite low as my plan is to separate reports per migration run, so i dont think performance will be an issue here.

and if it is, then we can swap the hashing algo here in the future.

graham-chainlink committed:
feat(operations): add hash cache on reports

Introduce a sync.map to store report cache computed from the sha256 to avoid repeat recomputation.
